runtime/include/QuantumDevice.hpp

```diff
@@ -55,310 +57,558 @@ struct QuantumDevice {
 
     /**
-     * @brief Compute partial samples with the number of shots on `wires`,
-     * returing raw samples.
-     *
-     * @param samples The pre-allocated `DataView<double, 2>`representing a matrix of
-     * shape `shots * numWires`. The built-in iterator in `DataView<double, 2>`
-     * iterates over all elements of `samples` row-wise.
-     * @param wires Wires to compute samples on
-     * @param shots The number of shots
+     * @brief (Optional) Start recording a quantum tape if provided.
+     *
+     * See `Gradient` for additional information.
      */
-    virtual void PartialSample(DataView<double, 2> &samples, const std::vector<QubitIdType> &wires,
-                               size_t shots) = 0;
+    virtual void StartTapeRecording() { RT_FAIL("Differentiation is unsupported by device"); }
 
     /**
-     * @brief Sample with the number of shots on the entire wires, returning the
-     * number of counts for each sample.
+     * @brief (Optional) Stop recording a quantum tape if provided.
      *
-     * @param eigvals The pre-allocated `DataView<double, 1>`
-     * @param counts The pre-allocated `DataView<int64_t, 1>`
-     * @param shots The number of shots
+     * See `Gradient` for additional information.
      */
-    virtual void Counts(DataView<double, 1> &eigvals, DataView<int64_t, 1> &counts,
-                        size_t shots) = 0;
+    virtual void StopTapeRecording() { RT_FAIL("Differentiation is unsupported by device"); }
+
+    // ----------------------------------------
+    //  DEVICE
+    // ----------------------------------------
 
     /**
-     * @brief Partial sample with the number of shots on `wires`, returning the
-     * number of counts for each sample.
+     * @brief Produce a measurement result value for "0" (computational basis).
      *
-     * @param eigvals The pre-allocated `DataView<double, 1>`
-     * @param counts The pre-allocated `DataView<int64_t, 1>`
-     * @param wires Wires to compute samples on
-     * @param shots The number of shots
+     * Results produced by projective qubit measurements (see `Measure`) have to be of the `Result`
+     * datatype. Currently, this type is defined as `bool*`, but in principle it could be a
+     * implementation-defined type supported by additional runtime functions. For the latter case,
+     * these functions exist to produce a known measurement value ("0" or "1") and compare it
+     * against another measurement value (like those produced by `Measure`). In the absence of such
+     * functionality however, these methods don't serve much of a purpose, but should be implemented
+     * nontheless for compliance with the interface.
+     *
+     * @return `Result` A "0" measurement result.
      */
-    virtual void PartialCounts(DataView<double, 1> &eigvals, DataView<int64_t, 1> &counts,
-                               const std::vector<QubitIdType> &wires, size_t shots) = 0;
+    [[nodiscard]] virtual auto Zero() const -> Result = 0;
```

Comment:
I would mention where `Result` is defined in the docstring.

```
See `Types.h` for additional details.
```

Reply:
There is a TODO proposing to remove custom data types in this interface entirely, please let me know what you think. In the meantime I will this comment.